findatastructures/src/finmap/mod.rs

```diff
@@ -0,0 +1,28 @@
+pub trait FinMap: PartialEq + Clone {
+    /// The type of keys in the map
+    type K;
+    /// The type of values in the map
+    type V;
+
+    /// Create a new empty map
+    fn new() -> Self;
+    /// Check if the map contains a key
+    fn contains_key(&self, k: &Self::K) -> bool;
+    /// Insert a key-value pair into the map
+    fn insert(&self, k: Self::K, v: Self::V) -> Self;
+    /// Get the value associated with a key
+    fn get(&self, k: &Self::K) -> Option<&Self::V>;
+    /// Remove a key-value pair from the map
+    fn remove(&self, k: &Self::K) -> Self;
+    /// Check if the map is empty
+    fn is_empty(&self) -> bool;
+    /// Get the number of key-value pairs in the map
+    fn len(&self) -> usize;
+    /// Get the keys in the map
+    fn keys(&self) -> Vec<&Self::K>;
```

Comment:
`keys` and `values` functions usually return an iterator. You could also use a slice if iterators are too complex. What is the design rational behind using a vector here?

Reply:
iterators are indeed complex, having to add lifetime variables to the trait and impl, seems like another battle with the borrowchecker.

I've also tried to look into slices, the issue is that the struct is a single vec of pairs, returning slices would require changing the struct to two parallel vecs one for keys another for values.

In light of this, should we change the struct? Or is Vec as a return type justified?